crates/bevy_math/src/smooth.rs

```diff
@@ -0,0 +1,207 @@
+use crate::{Vec2, Vec3};
+
+/// Smooths value to a goal using a damped spring.
+pub trait SmoothDamp {
+    /// Smooths value to a goal using a damped spring.
+    ///
+    /// `smooth_time` is the expected time to reach the target when at maximum velocity.
+    ///
+    /// Returns smoothed value and new velocity.
+    ///
+    /// # Panics
+    /// Panics if `smooth_time <= 0.0`.
+    ///
+    /// # Example
+    /// ```
+    /// # use bevy_math::prelude::{Vec3, Quat};
+    /// # use bevy_math::SmoothDamp;
+    /// # struct Transform {
+    /// #     translation: Vec3,
+    /// #     rotation: Quat,
+    /// #     scale: Vec3
+    /// # }
+    /// struct SmoothTransform {
+    ///     pub smoothness: f32,
+    ///     pub target: Vec3,   
+    ///     velocity: Vec3   
+    /// }
+    ///
+    /// fn smooth_transform_update(dt: f32, transform: &mut Transform, smoother: &mut SmoothTransform) {
+    ///     let (p, v) = Vec3::smooth_damp(
+    ///         transform.translation,
+    ///         smoother.target,
+    ///         smoother.velocity,
+    ///         smoother.smoothness,
+    ///         dt,
+    ///     );
+    ///     transform.translation = p;
+    ///     smoother.velocity = v;
+    ///     // When destructured assignement will be supported by Rust:
+    ///     // (transform.translation, smoother.velocity) =
+    ///     //     Vec3::smooth_damp(
+    ///     //         transform.translation,
+    ///     //         smoother.target,
+    ///     //         smoother.velocity,
+    ///     //         smoother.smoothness,
+    ///     //         dt,
+    ///     //      );
+    /// }
+    /// ```
+    fn smooth_damp(
+        from: Self,
+        to: Self,
+        velocity: Self,
+        smooth_time: f32,
+        delta_time: f32,
+    ) -> (Self, Self)
```

Comment:
I'm still not a fan of return a tuple and this is against Rust's ability to provide strong typing. 
I would prefer a 
```rust
struct SmoothDamp<T> {
    smoothed_value: T,
    velocity: T,
}
```

Reply:
I actually just learnt that struct destructuring, like tuple destructuring, is a thing in rust. So I'm going to give this another try :)